For a more complex data set it may need to be split multiple times to come up with the decision boundaries 
This is different in comparison to a simple data set where ridge regression can be used to draw when line for the decision boundaries
The order which the decsion tree is split up is important 
The order which the data sets attributes are selected will affect performance
To select the ordering system - 
You want an ordering to get a pure subset that gives the lowest entropy (the measure of randomness in the sample)
For example splitting the teams that average over a certain amount of points into groups. There may be a group that averages over a certain amount of points that always win their games. 1:3 or 6:2 win vs losses for the teams in hypothetical groups is relatively low entropy. This is low Entropy. Having a 50/50 split means there is a high entropy and the entopy is 1 total randomness. When teams are split up into groups based off of attributes you want the majority of the results to either be that the team wins or loses. 
With an ordering system that achieves low entropy there is a High Information Gain 
With an ordering system that achieves high entropy there is a Low Information Gain
Using an approach that leads to high information gain at every decision split 
Gini Impurity 
Where after splitting a sample ( putting a line at the midpoint for example there may be majority wins at the top and majority losses at the bottom but there might be one data point where ther is a win at the bottom) there wil be an anomalous data point. The decision split is almost pure but there is a little bit of impurity. Similar to entropy. Mathematics  


In [1]:
import pandas as pd 
games = pd.read_csv("nba_games.csv", index_col=0)
games = games.reset_index(drop = True)
games = games.sort_values("date")
games = games.reset_index(drop = True)
games

mp  mp.1  fg  fga    fg%  3p  3pa    3p%  ft  fta  ...  tov%_max_opp  \
0      240   240  33   87  0.379   5   16  0.313  22   29  ...          36.8   
1      240   240  43   92  0.467   8   26  0.308   9   10  ...          33.3   
2      240   240  36   91  0.396  15   33  0.455  17   18  ...          31.7   
3      240   240  31   84  0.369   6   19  0.316  13   17  ...          50.0   
4      240   240  34   85  0.400  10   32  0.313  16   18  ...          43.6   
...    ...   ...  ..  ...    ...  ..  ...    ...  ..  ...  ...           ...   
17767  240   240  43   97  0.443  16   38  0.421  19   21  ...          16.8   
17768  240   240  47   85  0.553  19   41  0.463  26   30  ...          26.2   
17769  240   240  36  102  0.353  12   42  0.286  22   27  ...          24.0   
17770  240   240  41  100  0.410  11   35  0.314  15   16  ...          37.5   
17771  240   240  35   86  0.407  12   27  0.444  26   30  ...          50.0   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
0              40.4           150           105       PHI         84   
1              44.9           200           120       ORL         91   
2              30.6           174           130       CHO         94   
3              42.7           195           102       CHI        108   
4              25.8           150           120       TOR        104   
...             ...           ...           ...       ...        ...   
17767          36.4           160           137       CLE        118   
17768          31.1           165           146       POR        106   
17769          46.7           286           117       LAL        139   
17770          34.3           178           117       BOS        123   
17771          50.2           217           120       IND        106   

       home_opp  season        date    won  
0             0    2016  01/01/2016   True  
1             0    2016  01/01/2016   True  
2             0    2016  01/01/2016   True  
3             1    2016  01/01/2016  False  
4             1    2016  01/01/2016  False  
...         ...     ...         ...    ...  
17767         1    2022  31/12/2021   True  
17768         0    2022  31/12/2021   True  
17769         1    2022  31/12/2021  False  
17770         1    2022  31/12/2021  False  
17771         1    2022  31/12/2021   True  

[17772 rows x 150 columns]

In [2]:
del games["mp.1"]
del games["mp_opp.1"]
del games["index_opp"]

In [3]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  tov%_max_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...          36.8   
1      240  43   92  0.467   8   26  0.308   9   10  0.900  ...          33.3   
2      240  36   91  0.396  15   33  0.455  17   18  0.944  ...          31.7   
3      240  31   84  0.369   6   19  0.316  13   17  0.765  ...          50.0   
4      240  34   85  0.400  10   32  0.313  16   18  0.889  ...          43.6   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  43   97  0.443  16   38  0.421  19   21  0.905  ...          16.8   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...          26.2   
17769  240  36  102  0.353  12   42  0.286  22   27  0.815  ...          24.0   
17770  240  41  100  0.410  11   35  0.314  15   16  0.938  ...          37.5   
17771  240  35   86  0.407  12   27  0.444  26   30  0.867  ...          50.0   

       usg%_max_opp  ortg_max_opp  drtg_max_opp  team_opp  total_opp  \
0              40.4           150           105       PHI         84   
1              44.9           200           120       ORL         91   
2              30.6           174           130       CHO         94   
3              42.7           195           102       CHI        108   
4              25.8           150           120       TOR        104   
...             ...           ...           ...       ...        ...   
17767          36.4           160           137       CLE        118   
17768          31.1           165           146       POR        106   
17769          46.7           286           117       LAL        139   
17770          34.3           178           117       BOS        123   
17771          50.2           217           120       IND        106   

       home_opp  season        date    won  
0             0    2016  01/01/2016   True  
1             0    2016  01/01/2016   True  
2             0    2016  01/01/2016   True  
3             1    2016  01/01/2016  False  
4             1    2016  01/01/2016  False  
...         ...     ...         ...    ...  
17767         1    2022  31/12/2021   True  
17768         0    2022  31/12/2021   True  
17769         1    2022  31/12/2021  False  
17770         1    2022  31/12/2021  False  
17771         1    2022  31/12/2021   True  

[17772 rows x 147 columns]

In [4]:
def get_target(team):
    team["target"] = team["won"].shift(-1)
    return team

games = games.groupby("team", group_keys=False).apply(get_target)

In [5]:
games[games["team"] =="LAL"]

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  usg%_max_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...          40.4   
14     240  37   92  0.402  15   32  0.469  25   31  0.806  ...          41.0   
25     240  34   76  0.447   8   26  0.308  20   29  0.690  ...          30.1   
40     240  43   90  0.478   7   28  0.250  24   34  0.706  ...          39.1   
65     240  41  101  0.406  14   33  0.424  13   16  0.813  ...          31.8   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17636  240  45   91  0.495  12   26  0.462  11   14  0.786  ...          28.8   
17647  240  43   84  0.512  10   17  0.588  18   24  0.750  ...          32.6   
17665  240  39   96  0.406   9   31  0.290   8   13  0.615  ...          28.4   
17697  290  53  105  0.505  10   25  0.400  26   37  0.703  ...          32.1   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...          31.1   

       ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  season  \
0               150           105       PHI         84         0    2016   
14              226           128       TOR        123         0    2017   
25              177           105       MIN        114         1    2018   
40              204           125       PHO        107         0    2020   
65              139           117       SAS        103         1    2021   
...             ...           ...       ...        ...       ...     ...   
17636           146           123       DET         93         0    2018   
17647           227           115       DAL        113         0    2019   
17665           208           103       HOU         85         0    2022   
17697           171           130       HOU        148         1    2018   
17768           165           146       POR        106         0    2022   

             date    won  target  
0      01/01/2016   True   False  
14     01/01/2017  False   False  
25     01/01/2018  False    True  
40     01/01/2020   True    True  
65     01/01/2021   True    True  
...           ...    ...     ...  
17636  31/10/2017   True    True  
17647  31/10/2018   True    True  
17665  31/10/2021   True   False  
17697  31/12/2017  False    True  
17768  31/12/2021   True     NaN  

[581 rows x 148 columns]

In [6]:
games["team"]

0        LAL
1        WAS
2        TOR
3        NYK
4        CHO
        ... 
17767    ATL
17768    LAL
17769    POR
17770    PHO
17771    CHI
Name: team, Length: 17772, dtype: object

In [7]:
games["target"][pd.isnull(games["target"])] = 2

C:\Users\Garnet\AppData\Local\Temp\ipykernel_16320\1495898042.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  games["target"][pd.isnull(games["target"])] = 2


In [8]:
games["target"] = games["target"].astype(int, errors="ignore")

In [9]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  usg%_max_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...          40.4   
1      240  43   92  0.467   8   26  0.308   9   10  0.900  ...          44.9   
2      240  36   91  0.396  15   33  0.455  17   18  0.944  ...          30.6   
3      240  31   84  0.369   6   19  0.316  13   17  0.765  ...          42.7   
4      240  34   85  0.400  10   32  0.313  16   18  0.889  ...          25.8   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  43   97  0.443  16   38  0.421  19   21  0.905  ...          36.4   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...          31.1   
17769  240  36  102  0.353  12   42  0.286  22   27  0.815  ...          46.7   
17770  240  41  100  0.410  11   35  0.314  15   16  0.938  ...          34.3   
17771  240  35   86  0.407  12   27  0.444  26   30  0.867  ...          50.2   

       ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  season  \
0               150           105       PHI         84         0    2016   
1               200           120       ORL         91         0    2016   
2               174           130       CHO         94         0    2016   
3               195           102       CHI        108         1    2016   
4               150           120       TOR        104         1    2016   
...             ...           ...       ...        ...       ...     ...   
17767           160           137       CLE        118         1    2022   
17768           165           146       POR        106         0    2022   
17769           286           117       LAL        139         1    2022   
17770           178           117       BOS        123         1    2022   
17771           217           120       IND        106         1    2022   

             date    won  target  
0      01/01/2016   True       0  
1      01/01/2016   True       0  
2      01/01/2016   True       1  
3      01/01/2016  False       0  
4      01/01/2016  False       0  
...           ...    ...     ...  
17767  31/12/2021   True       2  
17768  31/12/2021   True       2  
17769  31/12/2021  False       2  
17770  31/12/2021  False       2  
17771  31/12/2021   True       2  

[17772 rows x 148 columns]

In [10]:
games["won"].value_counts()

True     8886
False    8886
Name: won, dtype: int64

In [11]:
games["target"].value_counts()

1    8872
0    8870
2      30
Name: target, dtype: int64

In [12]:
nulls = pd.isnull(games)
nulls = nulls.sum()
nulls = nulls[nulls > 0]
nulls

+/-             17772
mp_max          17772
mp_max.1        17772
+/-_opp         17772
mp_max_opp      17772
mp_max_opp.1    17772
dtype: int64

In [13]:
valid_columns = games.columns[~games.columns.isin(nulls.index)]
valid_columns

Index(['mp', 'fg', 'fga', 'fg%', '3p', '3pa', '3p%', 'ft', 'fta', 'ft%',
       ...
       'usg%_max_opp', 'ortg_max_opp', 'drtg_max_opp', 'team_opp', 'total_opp',
       'home_opp', 'season', 'date', 'won', 'target'],
      dtype='object', length=142)

In [14]:
games = games[valid_columns].copy()
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  usg%_max_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...          40.4   
1      240  43   92  0.467   8   26  0.308   9   10  0.900  ...          44.9   
2      240  36   91  0.396  15   33  0.455  17   18  0.944  ...          30.6   
3      240  31   84  0.369   6   19  0.316  13   17  0.765  ...          42.7   
4      240  34   85  0.400  10   32  0.313  16   18  0.889  ...          25.8   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...           ...   
17767  240  43   97  0.443  16   38  0.421  19   21  0.905  ...          36.4   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...          31.1   
17769  240  36  102  0.353  12   42  0.286  22   27  0.815  ...          46.7   
17770  240  41  100  0.410  11   35  0.314  15   16  0.938  ...          34.3   
17771  240  35   86  0.407  12   27  0.444  26   30  0.867  ...          50.2   

       ortg_max_opp  drtg_max_opp  team_opp  total_opp  home_opp  season  \
0               150           105       PHI         84         0    2016   
1               200           120       ORL         91         0    2016   
2               174           130       CHO         94         0    2016   
3               195           102       CHI        108         1    2016   
4               150           120       TOR        104         1    2016   
...             ...           ...       ...        ...       ...     ...   
17767           160           137       CLE        118         1    2022   
17768           165           146       POR        106         0    2022   
17769           286           117       LAL        139         1    2022   
17770           178           117       BOS        123         1    2022   
17771           217           120       IND        106         1    2022   

             date    won  target  
0      01/01/2016   True       0  
1      01/01/2016   True       0  
2      01/01/2016   True       1  
3      01/01/2016  False       0  
4      01/01/2016  False       0  
...           ...    ...     ...  
17767  31/12/2021   True       2  
17768  31/12/2021   True       2  
17769  31/12/2021  False       2  
17770  31/12/2021  False       2  
17771  31/12/2021   True       2  

[17772 rows x 142 columns]

In [15]:
#splitting the data into just the targets column and all the other columns
#inputs = games.drop('target',axis='columns')
#target = games["target"]

In [16]:
#inputs

In [17]:
#target

In [18]:
games ["venue_code"] = games["home"].astype("category").cat.codes 

In [19]:
games ["opp_code"] = games["team_opp"].astype("category").cat.codes

In [20]:
games['date'] = pd.to_datetime(games['date'])

C:\Users\Garnet\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '13/01/2016' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\Garnet\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '13/01/2017' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\Garnet\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '13/01/2018' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\Garnet\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '13/01/2019' in DD/MM/YYYY format. Provide format or sp

In [21]:
games ["day_code"] = games ["date"].dt.dayofweek

In [22]:
games

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  team_opp  \
0      240  33   87  0.379   5   16  0.313  22   29  0.759  ...       PHI   
1      240  43   92  0.467   8   26  0.308   9   10  0.900  ...       ORL   
2      240  36   91  0.396  15   33  0.455  17   18  0.944  ...       CHO   
3      240  31   84  0.369   6   19  0.316  13   17  0.765  ...       CHI   
4      240  34   85  0.400  10   32  0.313  16   18  0.889  ...       TOR   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17767  240  43   97  0.443  16   38  0.421  19   21  0.905  ...       CLE   
17768  240  47   85  0.553  19   41  0.463  26   30  0.867  ...       POR   
17769  240  36  102  0.353  12   42  0.286  22   27  0.815  ...       LAL   
17770  240  41  100  0.410  11   35  0.314  15   16  0.938  ...       BOS   
17771  240  35   86  0.407  12   27  0.444  26   30  0.867  ...       IND   

       total_opp  home_opp  season       date    won  target  venue_code  \
0             84         0    2016 2016-01-01   True       0           1   
1             91         0    2016 2016-01-01   True       0           1   
2             94         0    2016 2016-01-01   True       1           1   
3            108         1    2016 2016-01-01  False       0           0   
4            104         1    2016 2016-01-01  False       0           0   
...          ...       ...     ...        ...    ...     ...         ...   
17767        118         1    2022 2021-12-31   True       2           0   
17768        106         0    2022 2021-12-31   True       2           1   
17769        139         1    2022 2021-12-31  False       2           0   
17770        123         1    2022 2021-12-31  False       2           0   
17771        106         1    2022 2021-12-31   True       2           0   

       opp_code  day_code  
0            22         4  
1            21         4  
2             4         4  
3             3         4  
4            27         4  
...         ...       ...  
17767         5         4  
17768        24         4  
17769        13         4  
17770         1         4  
17771        11         4  

[17772 rows x 145 columns]

In [23]:
#Due to machine learning algorithms like this decision tree only working on numbers and not Labels
#to turn the team and team_opp columns into numbers the LabelEncoder can be used 
from sklearn.preprocessing import LabelEncoder

In [24]:
#creating an object of the label encoder class for the column team
#team_code = LabelEncoder()
#creating an object of the label encoder class for the column team
#team_opp_code = LabelEncoder()

In [25]:
#creating new columns
#inputs['team_code'] = team_code.fit_transform(inputs['team'])
#inputs['team_opp_code'] = team_opp_code.fit_transform(inputs['team_opp'])
#inputs.head()

In [26]:
#inputs.head()

In [27]:
#dropping the columns with the team and opponent teams labels 
#inputs_n = inputs.drop(['team','team_opp','date'], axis='columns')
#inputs_n

In [28]:
#training the classifier 
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split 

In [29]:
clf = DecisionTreeClassifier(max_leaf_nodes=3, random_state=1)

In [30]:
#predictors = ["venue_code","opp_code","day_code"]

In [31]:
#the same data set for training and testing are used
#find out how to split up training and testing data set
#model.score(inputs_n,target)

In [32]:
#model.predict([[inputs]])

In [33]:
##X_train, X_test, y_trainy_test = train_test_split(inputs_n.drop('targets',axis=1), inputs_n['targets'], test_size=0.2)

In [34]:
#this splits the data into training and testing sets
predictors = ["venue_code","opp_code","day_code"]
X_train = games[games["date"]<'2022-01-01']
y_train = games[games["date"]<'2022-01-01']
X_test = games[games["date"]>'2022-01-01']
y_test = games[games["date"]>'2022-01-01']

In [35]:
clf.fit(X_train[predictors], y_train["target"])

DecisionTreeClassifier(max_leaf_nodes=3, random_state=1)

In [36]:
preds=clf.predict(X_test[predictors])

In [37]:
from sklearn.metrics import accuracy_score

In [38]:
acc = accuracy_score(X_test["target"], preds)

In [39]:
acc

0.5006329113924051

In [40]:
combined = pd.DataFrame(dict(actual=X_test["target"], prediction=preds))

In [41]:
pd.crosstab(index= combined["actual"],columns=combined["prediction"])

prediction    0    1
actual              
0           405  404
1           385  386

405 times the model was correct and 404 times the model was wrong
386 times the model was correct and 385 times the model was wrong

In [42]:
from sklearn.metrics import precision_score 

In [43]:
precision_score(X_test["target"], preds)

0.48860759493670886

When a win was predicted the team won 48.86% of the time which is below 50% 

# Improvement of the model 
##### Rolling averages by splitting the data set up into team 
###### The folloing data groups the data set into team statistics

In [44]:
grouped_games = games.groupby("team")
group = grouped_games.get_group("GSW")
group

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  team_opp  \
54     240  31   79  0.392   7   35  0.200  29   33  0.879  ...       POR   
76     240  46   86  0.535  17   32  0.531  14   21  0.667  ...       UTA   
97     240  47   96  0.490  21   48  0.438  11   14  0.786  ...       CHO   
147    240  48   90  0.533  18   35  0.514  17   19  0.895  ...       CLE   
187    240  43   96  0.448  20   46  0.435  18   18  1.000  ...       SAS   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17621  240  46   84  0.548  17   30  0.567  25   35  0.714  ...       NOP   
17642  240  47   89  0.528  16   32  0.500  21   30  0.700  ...       NOP   
17679  240  43   84  0.512  11   23  0.478  17   27  0.630  ...       HOU   
17712  240  46   84  0.548  14   28  0.500  26   32  0.813  ...       PHO   
17727  265  44   96  0.458  12   27  0.444  13   18  0.722  ...       SAS   

       total_opp  home_opp  season       date    won  target  venue_code  \
54           123         0    2021 2021-01-01  False       1           1   
76           116         1    2022 2022-01-01   True       1           0   
97           111         0    2017 2017-01-02   True       1           1   
147          112         1    2020 2020-01-02   True       1           0   
187          120         1    2022 2022-01-02   True       1           0   
...          ...       ...     ...        ...    ...     ...         ...   
17621        120         1    2016 2015-10-31   True       1           0   
17642        121         0    2019 2018-10-31   True       1           1   
17679        110         1    2016 2015-12-31   True       1           0   
17712        109         1    2019 2018-12-31   True       0           0   
17727        117         1    2020 2019-12-31  False       2           0   

       opp_code  day_code  
54           24         4  
76           28         5  
97            4         0  
147           5         3  
187          26         6  
...         ...       ...  
17621        18         5  
17642        18         2  
17679        10         3  
17712        23         0  
17727        26         1  

[655 rows x 145 columns]

The games of GSW (Golden State Warriors)if the prediction is on the fourth game the model should look at the previous three games to help make better predictions. For example if the team is on a losing streak or a winning streak

In [45]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    #closed = ''left' only have the rolling averages of the previous weeks not the week we are trying to calculate for
    rolling_stats = group[cols].rolling(3,closed='left').mean()
    group[new_cols] = rolling_stats
    #removes all of the rows that have missing values 
    group = group.dropna(subset= new_cols)
    return group

Using the same roling averages as the ones used for Random Forest

In [49]:
#The same cols are used for the rolling average to keep the inputs consistent 
cols = ["fg","3p","ft","ast","stl","blk","tov", "pts"]
new_cols = [f"{c}_rolling" for c in cols]

In [50]:
new_cols

['fg_rolling',
 '3p_rolling',
 'ft_rolling',
 'ast_rolling',
 'stl_rolling',
 'blk_rolling',
 'tov_rolling',
 'pts_rolling']

In [51]:
rolling_averages(group, cols, new_cols)

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  opp_code  \
3100  240  35   73  0.479  16   30  0.533  26   36  0.722  ...        27   
3672  240  46   93  0.495  17   37  0.459  10   18  0.556  ...         7   
3735  240  44   86  0.512  14   33  0.424  12   23  0.522  ...         2   
4229  240  38   92  0.413   8   39  0.205  19   22  0.864  ...        25   
4953  240  52   95  0.547  16   35  0.457  11   15  0.733  ...        11   
...   ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
6097  240  40   91  0.440  15   43  0.349  12   15  0.800  ...         1   
6372  240  35   82  0.427  13   36  0.361  25   29  0.862  ...        14   
6697  240  36   80  0.450  14   39  0.359   9   13  0.692  ...        14   
7086  240  40   90  0.444  12   38  0.316  25   34  0.735  ...        13   
7173  240  45   93  0.484  18   45  0.400  14   15  0.933  ...        16   

      day_code  fg_rolling  3p_rolling  ft_rolling  ast_rolling  stl_rolling  \
3100         1   42.666667   12.333333   18.000000    30.333333     5.333333   
3672         3   40.000000   14.000000   19.333333    27.000000     3.333333   
3735         4   40.000000   14.333333   20.000000    30.000000     6.000000   
4229         5   41.666667   15.666667   16.000000    29.000000     7.000000   
4953         2   42.666667   13.000000   13.666667    30.666667    10.666667   
...        ...         ...         ...         ...          ...          ...   
6097         3   44.333333   16.000000   13.666667    28.666667     6.000000   
6372         1   43.666667   16.000000   12.666667    26.333333     8.000000   
6697         5   41.333333   15.666667   16.000000    26.000000     7.666667   
7086         4   37.000000   14.000000   15.333333    22.333333     7.000000   
7173         5   37.000000   13.000000   19.666667    23.666667     5.000000   

      blk_rolling  tov_rolling  pts_rolling  
3100     9.000000    14.666667   115.666667  
3672     6.333333    14.000000   113.333333  
3735     5.333333    14.666667   114.333333  
4229     6.666667    14.666667   115.000000  
4953     5.666667    16.666667   112.000000  
...           ...          ...          ...  
6097     3.666667    10.333333   118.333333  
6372     3.666667    13.333333   116.000000  
6697     4.000000    16.333333   114.333333  
7086     5.000000    18.000000   103.333333  
7173     4.333333    15.666667   106.666667  

[652 rows x 153 columns]

After getting the original nba_games data Frame and grouped by teams which creates one data frame for each team in the data. Then the lambda function was applied to compute the rolling averages.

In [52]:
games_rolling = games.groupby("team").apply(lambda x:rolling_averages(x, cols, new_cols)) 

###### Rolling averages are taken from the entire data set

In [53]:
games_rolling

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  opp_code  \
team                                                            ...             
ATL  2383  240  37   76  0.487   8   22  0.364  19   24  0.792  ...         2   
     2462  240  37   74  0.500  10   26  0.385  16   20  0.800  ...        13   
     3659  240  41   77  0.532   9   21  0.429  30   35  0.857  ...        18   
     4228  240  43   92  0.467  13   33  0.394  15   15  1.000  ...        29   
     5486  240  44   86  0.512  13   31  0.419   6    7  0.857  ...        17   
...        ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
WAS  6378  240  47   87  0.540  11   33  0.333  17   23  0.739  ...        20   
     6587  240  39   85  0.459   9   27  0.333  22   33  0.667  ...        13   
     6991  240  43   82  0.524  10   31  0.323  16   21  0.762  ...        21   
     7087  240  41   88  0.466  13   38  0.342  15   21  0.714  ...        25   
     7179  240  39   83  0.470  14   34  0.412  26   27  0.963  ...        24   

           day_code  fg_rolling  3p_rolling  ft_rolling  ast_rolling  \
team                                                                   
ATL  2383         5   35.666667    7.333333   14.000000    23.000000   
     2462         6   35.666667    7.666667   16.000000    24.666667   
     3659         3   37.000000    8.333333   17.333333    25.333333   
     4228         5   38.333333    9.000000   21.666667    27.000000   
     5486         4   40.333333   10.666667   20.333333    30.333333   
...             ...         ...         ...         ...          ...   
WAS  6378         1   42.000000    9.666667   16.333333    26.666667   
     6587         3   44.000000   10.000000   16.333333    27.666667   
     6991         3   42.666667    9.333333   18.333333    25.000000   
     7087         4   43.000000   10.000000   18.333333    25.333333   
     7179         5   41.000000   10.666667   17.666667    24.333333   

           stl_rolling  blk_rolling  tov_rolling  pts_rolling  
team                                                           
ATL  2383     9.000000     4.666667    12.666667    92.666667  
     2462    11.000000     6.000000    13.333333    95.000000  
     3659    12.000000     6.666667    15.000000    99.666667  
     4228    12.000000     8.333333    15.000000   107.333333  
     5486    11.000000     6.333333    15.333333   111.666667  
...                ...          ...          ...          ...  
WAS  6378     9.666667     3.666667    11.000000   110.000000  
     6587     7.333333     5.000000     9.666667   114.333333  
     6991     7.000000     5.333333     9.333333   113.000000  
     7087     5.000000     5.666667     9.333333   114.333333  
     7179     7.666667     4.333333     9.666667   110.333333  

[17682 rows x 153 columns]

Dropping the team name as a seperate level in the pandas index. for a multi level index you would need to use two values to access each row which was not needed

In [54]:
games_rolling = games_rolling.droplevel('team')

In [55]:
games_rolling 

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  opp_code  \
2383  240  37   76  0.487   8   22  0.364  19   24  0.792  ...         2   
2462  240  37   74  0.500  10   26  0.385  16   20  0.800  ...        13   
3659  240  41   77  0.532   9   21  0.429  30   35  0.857  ...        18   
4228  240  43   92  0.467  13   33  0.394  15   15  1.000  ...        29   
5486  240  44   86  0.512  13   31  0.419   6    7  0.857  ...        17   
...   ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
6378  240  47   87  0.540  11   33  0.333  17   23  0.739  ...        20   
6587  240  39   85  0.459   9   27  0.333  22   33  0.667  ...        13   
6991  240  43   82  0.524  10   31  0.323  16   21  0.762  ...        21   
7087  240  41   88  0.466  13   38  0.342  15   21  0.714  ...        25   
7179  240  39   83  0.470  14   34  0.412  26   27  0.963  ...        24   

      day_code  fg_rolling  3p_rolling  ft_rolling  ast_rolling  stl_rolling  \
2383         5   35.666667    7.333333   14.000000    23.000000     9.000000   
2462         6   35.666667    7.666667   16.000000    24.666667    11.000000   
3659         3   37.000000    8.333333   17.333333    25.333333    12.000000   
4228         5   38.333333    9.000000   21.666667    27.000000    12.000000   
5486         4   40.333333   10.666667   20.333333    30.333333    11.000000   
...        ...         ...         ...         ...          ...          ...   
6378         1   42.000000    9.666667   16.333333    26.666667     9.666667   
6587         3   44.000000   10.000000   16.333333    27.666667     7.333333   
6991         3   42.666667    9.333333   18.333333    25.000000     7.000000   
7087         4   43.000000   10.000000   18.333333    25.333333     5.000000   
7179         5   41.000000   10.666667   17.666667    24.333333     7.666667   

      blk_rolling  tov_rolling  pts_rolling  
2383     4.666667    12.666667    92.666667  
2462     6.000000    13.333333    95.000000  
3659     6.666667    15.000000    99.666667  
4228     8.333333    15.000000   107.333333  
5486     6.333333    15.333333   111.666667  
...           ...          ...          ...  
6378     3.666667    11.000000   110.000000  
6587     5.000000     9.666667   114.333333  
6991     5.333333     9.333333   113.000000  
7087     5.666667     9.333333   114.333333  
7179     4.333333     9.666667   110.333333  

[17682 rows x 153 columns]

In [56]:
games_rolling.index = range(games_rolling.shape[0])

In [57]:
games_rolling

mp  fg  fga    fg%  3p  3pa    3p%  ft  fta    ft%  ...  opp_code  \
0      240  37   76  0.487   8   22  0.364  19   24  0.792  ...         2   
1      240  37   74  0.500  10   26  0.385  16   20  0.800  ...        13   
2      240  41   77  0.532   9   21  0.429  30   35  0.857  ...        18   
3      240  43   92  0.467  13   33  0.394  15   15  1.000  ...        29   
4      240  44   86  0.512  13   31  0.419   6    7  0.857  ...        17   
...    ...  ..  ...    ...  ..  ...    ...  ..  ...    ...  ...       ...   
17677  240  47   87  0.540  11   33  0.333  17   23  0.739  ...        20   
17678  240  39   85  0.459   9   27  0.333  22   33  0.667  ...        13   
17679  240  43   82  0.524  10   31  0.323  16   21  0.762  ...        21   
17680  240  41   88  0.466  13   38  0.342  15   21  0.714  ...        25   
17681  240  39   83  0.470  14   34  0.412  26   27  0.963  ...        24   

       day_code  fg_rolling  3p_rolling  ft_rolling  ast_rolling  stl_rolling  \
0             5   35.666667    7.333333   14.000000    23.000000     9.000000   
1             6   35.666667    7.666667   16.000000    24.666667    11.000000   
2             3   37.000000    8.333333   17.333333    25.333333    12.000000   
3             5   38.333333    9.000000   21.666667    27.000000    12.000000   
4             4   40.333333   10.666667   20.333333    30.333333    11.000000   
...         ...         ...         ...         ...          ...          ...   
17677         1   42.000000    9.666667   16.333333    26.666667     9.666667   
17678         3   44.000000   10.000000   16.333333    27.666667     7.333333   
17679         3   42.666667    9.333333   18.333333    25.000000     7.000000   
17680         4   43.000000   10.000000   18.333333    25.333333     5.000000   
17681         5   41.000000   10.666667   17.666667    24.333333     7.666667   

       blk_rolling  tov_rolling  pts_rolling  
0         4.666667    12.666667    92.666667  
1         6.000000    13.333333    95.000000  
2         6.666667    15.000000    99.666667  
3         8.333333    15.000000   107.333333  
4         6.333333    15.333333   111.666667  
...            ...          ...          ...  
17677     3.666667    11.000000   110.000000  
17678     5.000000     9.666667   114.333333  
17679     5.333333     9.333333   113.000000  
17680     5.666667     9.333333   114.333333  
17681     4.333333     9.666667   110.333333  

[17682 rows x 153 columns]

In [58]:
def make_predictions(data,predictors):
    X_train = data[data["date"]<'2022-01-01']
    y_test = data[data["date"]>'2022-01-01']
    clf.fit(X_train[predictors], X_train["target"])
    preds = clf.predict(y_test[predictors])
    combined = pd.DataFrame(dict(actual=y_test["target"], predicted=preds), index=y_test.index)
    precision = precision_score(y_test["target"], preds)
    return combined, precision 

In [59]:
combines, precision = make_predictions(games_rolling, predictors + new_cols)

In [60]:
precision

0.48056537102473496

The precision was 0.48860759493670886 now the precision is 0.48056537102473496 a decrease of 0.00804222391197390

In [61]:
combined

actual  prediction
178         0           0
179         0           1
180         0           0
181         1           1
182         0           1
...       ...         ...
17585       1           0
17586       1           0
17587       1           1
17588       0           1
17589       0           0

[1580 rows x 2 columns]

In [62]:
combined = combined.merge(games_rolling[["date", "team", "team_opp", "won"]], left_index=True, right_index=True)

In [63]:
combined

actual  prediction       date team team_opp    won
178         0           0 2017-09-12  ATL      ORL   True
179         0           1 2017-10-01  ATL      BRK   True
180         0           0 2017-10-02  ATL      SAC  False
181         1           1 2017-10-03  ATL      TOR   True
182         0           1 2017-10-11  ATL      DET  False
...       ...         ...        ...  ...      ...    ...
17585       1           0 2021-05-26  WAS      PHI  False
17586       1           0 2021-05-29  WAS      PHI  False
17587       1           1 2021-05-31  WAS      PHI   True
17588       0           1 2021-06-01  WAS      PHI  False
17589       0           0 2021-06-05  WAS      TOR   True

[1580 rows x 6 columns]

In [64]:
#to make sure the names of the home and away team are the same 
#the Dataset already has the same names for home and away teams
#for future reference if names need to be normalised
class MissingDict(dict):
    __missing__ = lambda self, key: key
    
map_values = {
    
    "Golden State Warriors": "GSW",
    "Lakers": "LAL",
    "Boston Celtics": "BOS"
    
}
mapping = MissingDict(**map_values)

In [65]:
mapping["Golden State Warriors"]

'GSW'

In [66]:
combined["new_team"] = combined["team"].map(mapping)

In [67]:
combined

actual  prediction       date team team_opp    won new_team
178         0           0 2017-09-12  ATL      ORL   True      ATL
179         0           1 2017-10-01  ATL      BRK   True      ATL
180         0           0 2017-10-02  ATL      SAC  False      ATL
181         1           1 2017-10-03  ATL      TOR   True      ATL
182         0           1 2017-10-11  ATL      DET  False      ATL
...       ...         ...        ...  ...      ...    ...      ...
17585       1           0 2021-05-26  WAS      PHI  False      WAS
17586       1           0 2021-05-29  WAS      PHI  False      WAS
17587       1           1 2021-05-31  WAS      PHI   True      WAS
17588       0           1 2021-06-01  WAS      PHI  False      WAS
17589       0           0 2021-06-05  WAS      TOR   True      WAS

[1580 rows x 7 columns]

This makes sure that the predictions are the same on both sides 

In [70]:

merged = combined.merge(combined, left_on=["date","new_team"], right_on=["date","team_opp"])

In [71]:
merged

actual_x  prediction_x       date team_x team_opp_x  won_x new_team_x  \
0           1             1 2018-11-25    ATL        CHO   True        ATL   
1           0             0 2018-11-28    ATL        CHO  False        ATL   
2           0             0 2020-01-14    ATL        PHO   True        ATL   
3           0             1 2020-01-26    ATL        WAS   True        ATL   
4           0             1 2020-02-22    ATL        DAL   True        ATL   
..        ...           ...        ...    ...        ...    ...        ...   
157         0             0 2021-02-02    UTA        DET   True        UTA   
158         1             0 2020-01-26    WAS        ATL  False        WAS   
159         1             0 2020-01-30    WAS        CHO   True        WAS   
160         1             1 2021-05-16    WAS        CHO   True        WAS   
161         0             0 2021-06-05    WAS        TOR   True        WAS   

     actual_y  prediction_y team_y team_opp_y  won_y new_team_y  
0           0             1    CHO        ATL  False        CHO  
1           0             1    CHO        ATL   True        CHO  
2           1             0    PHO        ATL  False        PHO  
3           1             0    WAS        ATL  False        WAS  
4           0             0    DAL        ATL  False        DAL  
..        ...           ...    ...        ...    ...        ...  
157         0             1    DET        UTA  False        DET  
158         0             1    ATL        WAS   True        ATL  
159         0             0    CHO        WAS  False        CHO  
160         0             1    CHO        WAS  False        CHO  
161         0             0    TOR        WAS  False        TOR  

[162 rows x 13 columns]

In [74]:
merged[(merged["prediction_x"] == 1) & (merged["prediction_y"] == 0)]["actual_x"].value_counts()

0    20
1    15
Name: actual_x, dtype: int64

In [75]:
15/20

0.75

The accuracy is 75% from merging both sides of the match together 
Using more seasons could improve accuracy
Using more columns in the data frame can improve accuracy 
Look at both current team and opponent team records for the season